### PR TITLE
Thomas scraper

### PR DESCRIPTION
**Pull request by t6george**
Aight this should work for the Bulbapedia Pokedex.

This ended up getting mossed because of the many types of Pokemon that exist under one name (for instance, one article talks about Charizard, Mega Charizard X, Mega Charizard Y and Gigantamax Charizard since they are all technically the same Pokemon but they look different).

Because of this I needed to make a state machine that parses the Biology section of each article and associates each paragraph to the appropriate version of the Pokemon that it refers to. It may not work perfectly, but it should work the vast majority of the time. However if an image cannot be matched with a paragraph/sequence of paragraphs, that data point should end up being discarded.

### CODE REVIEW

**Comment by q349wang** [comment]
Add new lines to end of files

**Comment by q349wang** [comment]
It would be good to start a requirements.txt or something for all the pip packages used

File: web/scraper/EntityScrapers/PokemonEntityScraper.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            megaXYEvolution = 'Mega ' + self.monName
+            del stateTransitions[ParseStates.Mega]
+            stateTransitions[ParseStates.MegaX] = re.compile(r'.*(%s).*' %(megaXYEvolution + ' X'), flags = re.IGNORECASE)
+            stateTransitions[ParseStates.MegaY] = re.compile(r'.*(%s).*' %(megaXYEvolution + ' Y'), flags = re.IGNORECASE)
+        
+        if not any(re.match(stateTransitions[ParseStates.Galarian], title) for title in self.imageTitles):
+            del stateTransitions[ParseStates.Galarian]
+
+        if not any(re.match(stateTransitions[ParseStates.Alolan], title) for title in self.imageTitles):
+            del stateTransitions[ParseStates.Alolan]
```
Comment (q349wang) [inline web/scraper/EntityScrapers/PokemonEntityScraper.py:92]: Honestly have no clue what this moss is. But I'll trust its just parsing

File: web/scraper/EntityScrapers/PokemonEntityScraper.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                    stateTransitions[ParseStates.Custom2] = re.compile(r'.*(%s).*' %title, flags = re.IGNORECASE)
+                    break
+                else:
+                    stateTransitions[ParseStates.Custom1] = re.compile(r'.*(%s).*' %title, flags = re.IGNORECASE)
+                    customStateAdded = True
+
+        naturalLanguage = self.getNaturalLanguage(monBiology, stateTransitions)
+
+        return [DescribedEntity(imageLinks[i], naturalLanguage[i]) \
+            for i in range(min(len(imageLinks), len(naturalLanguage)))]
```
Comment (q349wang) [inline web/scraper/EntityScrapers/PokemonEntityScraper.py:107]: are image urls and natural language guaranteed to be in the same order? What if the lengths are different. Should we use blank natural language if its not found?

File: web/scraper/EntityScrapers/PokemonEntityScraper.py (hunk 1 above)
Comment (t6george) [reply web/scraper/EntityScrapers/PokemonEntityScraper.py:92]: This handles the case where a Pokemon region such as the Galar Region is mentioned in the biology but the Pokemon does not have a special Galarian form. In other words, if there is no Galarian picture for a pokemon, then ignore the word 'Galarian' if it comes up in the text and treat it as a regular word. Same goes for Alolan Region.

File: web/scraper/EntityScrapers/PokemonEntityScraper.py (hunk 2 above)
Comment (t6george) [reply web/scraper/EntityScrapers/PokemonEntityScraper.py:107]: The images appear in the order of the paragraphs.

By using min(len(imageLinks), len(naturalLanguage)) we make sure that we dont run off the end of one of the lists.

In the odd case that the imageLinks and naturalLanguage lists are not the same length, the remainder of the unmatched data is discarded.

Comment (q349wang) [reply web/scraper/EntityScrapers/PokemonEntityScraper.py:107]: Ok got it, so no unlabeled data then?

Comment (t6george) [reply web/scraper/EntityScrapers/PokemonEntityScraper.py:107]: No unlabelled data. We drop any text with no associated image and any image with no associated text.

Comment (q349wang) [reply web/scraper/EntityScrapers/PokemonEntityScraper.py:107]: Sounds good

**Comment by q349wang** [comment]
test comment

File: web/scraper/ScraperList.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from EntityScrapers.PokemonEntityScraper import *
+from UrlListBuilders.PokemonUrlListBuilder import *
```
Comment (q349wang) [inline web/scraper/ScraperList.py:2]: test comment 2

**Review by q349wang — CHANGES_REQUESTED** [review]
Add the _pycache_ and .pyc files to a gitignore. Dont think we need them

**Review by q349wang — APPROVED** [review]

**Pull request merged**